flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/Checkpoints.java

```diff
@@ -369,4 +373,37 @@ public static CheckpointStorage loadCheckpointStorage(
 
     /** This class contains only static utility methods and is not meant to be instantiated. */
     private Checkpoints() {}
+
+    private static class ClaimModeCompletedStorageLocation
+            implements CompletedCheckpointStorageLocation {
+
+        private final CompletedCheckpointStorageLocation wrapped;
+
+        private ClaimModeCompletedStorageLocation(CompletedCheckpointStorageLocation location) {
+            wrapped = location;
+        }
+
+        @Override
+        public String getExternalPointer() {
+            return wrapped.getExternalPointer();
+        }
+
+        @Override
+        public StreamStateHandle getMetadataHandle() {
+            return wrapped.getMetadataHandle();
+        }
+
+        @Override
+        public void disposeStorageLocation() throws IOException {
+            try {
+                wrapped.disposeStorageLocation();
+            } catch (Exception ex) {
+                LOG.debug(
+                        "We could not delete the storage location: {} in CLAIM restore mode. It is"
+                                + " most probably because of shared files still being used by newer"
+                                + " checkpoints",
```

Comment:
@dawidwys could you please explain this scenario?

AFAIK, shared state files should be placed in a separate `checkpoints/shared` folder and therefore should not prevent `checkpoints/chk-xxx` folder from being deleted.
However, this is not true when migrating to Changelog, when private state becomes "re-usable".

I'd like to understand whether it's a general case or only related to Changelog.

cc: @zoltar9264

Reply:
This is the case for native savepoints. Native "incremental" RocksDB savepoints place all files in a single savepoint folder. They do not place shared files in the shared folder and thus they can prevent deleting the savepoint folder.

Reply:
Got it. Thanks for the explanation.

Reply:
So basically the check inside the FS implementation (that the folder is not empty) prevents savepoints from being invalidated. This seems a bit fragile, doesn't it?
Wouldn't it make more sense to use checkpoint properties to prevent `delete` call alltogether?

Reply:
It does not prevent savepoint from being invalidated. It simply prevents logging an error.